core/drivers/crypto/caam/caam_jr.c

```diff
@@ -0,0 +1,781 @@
+
+	if (nbJobs_done == 0) {
+		cpu_spin_unlock_xrestore(&jr_privdata->inlock, exceptions);
+		return retJobId;
+	}
+
+	/* Ensure that output ring descriptor entries are not in cache */
+	if ((jr_privdata->outread_index + nbJobs_done) > jr_privdata->nbJobs) {
+		/* Invalidate all job buffer */
+		/* Some job completed are at before the current outread_index */
+		jr_out     = jr_privdata->outrings;
+		nbJobs_inv = jr_privdata->nbJobs;
+	} else {
+		/* Invalidate only the job index completed */
+		jr_out = &jr_privdata->outrings[jr_privdata->outread_index];
+		nbJobs_inv = nbJobs_done;
+	}
+
+	cache_operation(TEE_CACHEINVALIDATE, (void *)jr_out,
+			(sizeof(struct outring_entry) * nbJobs_inv));
+
+	do {
+		jr_out = &jr_privdata->outrings[jr_privdata->outread_index];
+
+		/*
+		 * Lock the caller information array because enqueue is
+		 * also touching it
+		 */
+		cpu_spin_lock(&jr_privdata->callers_lock);
+		for ((idx_jr = 0), (found = (-1));
+				(idx_jr < jr_privdata->nbJobs) && (found < 0);
+				idx_jr++) {
+			/*
+			 * Search for the caller information corresponding to
+			 * the completed JR.
+			 * Don't use the outread_index or inwrite_index because
+			 * completion can be out of order compared to input
+			 * buffer
+			 */
+			caller = &jr_privdata->callers[idx_jr];
+			if (caam_read_val(&jr_out->desc) == caller->pdesc) {
+				jobctx         = caller->jobctx;
+				jobctx->status = caam_read_val(&jr_out->status);
+
+				/* Update return Job Id mask */
+				if (caller->jobid & waitJobIds)
+					retJobId |= caller->jobid;
+
+				JR_TRACE("JR id=%d, context @0x%08"PRIxVA"",
+					caller->jobid, (vaddr_t)jobctx);
+				/* Clear the Entry Descriptor DMA */
+				caller->pdesc = 0;
+				caller->jobid = JR_JOB_FREE;
+				found = idx_jr;
+				JR_TRACE("Free space #%d in the callers array",
+					idx_jr);
+			}
+		}
+		cpu_spin_unlock(&jr_privdata->callers_lock);
+
+		/*
+		 * Remove the JR from the output list even if no
+		 * JR caller found
+		 */
+		hal_jr_del_job(jr_privdata->baseaddr);
+
+		/*
+		 * Increment index to next JR output entry taking care that
+		 * is a circular buffer of nbJobs size.
+		 */
+		jr_privdata->outread_index++;
+		jr_privdata->outread_index %= jr_privdata->nbJobs;
+
+		if ((found >= 0) && (jobctx->callbk)) {
+			/* Finally, execute user's callback */
+			jobctx->callbk(jobctx);
+		}
+
+	} while (--nbJobs_done);
+
+	cpu_spin_unlock_xrestore(&jr_privdata->inlock, exceptions);
+
+	return retJobId;
+}
+
+/*
+ * Enqueues a new job in the Job Ring input queue. Keep the caller's
+ * job context in private array.
+ *
+ * Inputs:
+ * jobctx   Caller's job context
+ *
+ * Outputs:
+ * jobId    Job Id enqueued
+ *
+ * Returns:
+ * CAAM_NO_ERROR  Success
+ * CAAM_BUSY      CAAM is busy
+ */
+static enum CAAM_Status do_jr_enqueue(struct jr_jobctx *jobctx, uint32_t *jobId)
+{
+	enum CAAM_Status retstatus = CAAM_BUSY;
+
+	struct caller_info *caller = NULL;
+	uint32_t exceptions = 0;
+	uint32_t job_mask = 0;
+	uint8_t  idx_jr = 0;
```

Comment:
I'd prefer `uint8_t idx_jr = 0;` without the extra space, same on the row below.